Write Releases history to GCS

mozbhearsum opened the pull request:
This change gets Balrog writing Releases history to GCS on the backend, and reading it from GCS on the frontend. (It also needs to read it on the backend to deal with update races, but I've removed all of the endpoints that expose Releases history, since clients can get it direct from GCS.)

This code is running on the new GCS Balrog stage environment (https://admin-stage.balrog.nonprod.cloudops.mozgcp.net/releases), so feel free to play around there as much as you'd like. There's numerous Firefox-68.0b1 releases with history written to GCS. I can also give you credentials to my test bucket if you'd like to play around locally.

I have a separate patch in the works to import existing history into GCS. We'll want that landed first, so we can migrate the data before dropping the table. I hope to have that ready today or Monday.

If the patch looks like a mess when looking at is a whole it might be easier to look at it commit-by-commit, as I've rebased into pretty reasonable chunks. https://github.com/mozilla/balrog/pull/910/commits/fff9eab94259bc56f6dc2dd6e99c2b447b47997e and https://github.com/mozilla/balrog/pull/910/commits/f8f1b1055795c6c059be12857408fd79fd47fdae are the most interesting parts, which update the backend and frontend to use GCS. The rest is code removal, rearchitecting, and test changes.

mozbhearsum commented on ui/app/js/controllers/releases_controller.js:
```diff
@@ -64,8 +56,8 @@ function($scope, $routeParams, $location, $timeout, Releases, Search, $modal, Pa
-        text: "Data Version",
-        value: "-data_version"
+        text: "Timestamp",
+        value: "-timestamp"
```
While testing, I ended up doing deletes/inserts of the same Release a bunch of times, and discovered that we end up with duplicate data versions (even with the current code). Sorting by timestamp here improves the situation by actually showing you the changes in the order that they were made, rather than relegating both inserts (which have data_version=1) at the very last page.

nthomas-mozilla replied:
Very nice! 👍

mozbhearsum commented:
Because I was bound to forget something important: while reviewing and updating https://github.com/mozilla-releng/releng-rfcs/pull/21, I remembered that we actually need two separate buckets for this, because GCS Lifecycle Policies don't allow for configuration by key name or prefix.

nthomas-mozilla commented on ui/app/js/app.run.js:
```diff
@@ -6,3 +6,18 @@ angular.module("app").run(function($location, $http, Auth0) {
 });
+
+angular.module("app").factory("GCloudHttpInterceptor", function() {
+    return {
```
Only skimmed the ui/ changes.

mozbhearsum commented:
> Sorry for the delay reviewing this change, is big!

No worries!

> I get a 403 response from this but would like to have a play with it live. Could you take a look ?

I suspect you weren't connected to the VPN. For the new GCP environments, this is the expected response (unlike AWS, where we didn't return any response at all).

mozbhearsum commented on auslib/db.py:
```diff
@@ -805,107 +857,6 @@ def getChange(self, change_id=None, column_values=None, data_version=None, trans
             return None
         return changes[0]
 
-    def getPrevChange(self, change_id, row_primary_keys, transaction=None):
```
Ah, I should've added a note about this. As far as I can tell, `rollbackChange` was never actually used. It was added in https://bugzilla.mozilla.org/show_bug.cgi?id=747129, but the current UI has always made a `POST` to `/releases/:release/revisions` to revert, which has always used `Releases.update` to do the work.

So, this is just long overdue dead code removal that I noticed while auditing what `GCSHistory` needs to implement.

As far as more changes to the UI goes, you're right - I need to update the revert controller/service to use a `POST` to `/releases/:release` instead of the `revisions` endpoint.

nthomas-mozilla replied:
Ah, that makes sense. There is the 'Revert to this' UI on the rules history to check on, not sure what that code that uses.

*gentle nudge about the coverage comment above*

mozbhearsum replied:
> There is the 'Revert to this' UI on the rules history to check on, not sure what that code that uses.

That uses a POST to `/api/rules/:rule_id/revisions` which eventually ends up doing a direct `update` at https://github.com/mozilla/balrog/blob/master/auslib/web/admin/views/history.py#L136

> gentle nudge about the coverage comment above

Thanks, I'll address the coverage in a separate PR.

mozbhearsum commented on auslib/web/public/swagger/public_api_spec.yml:
```diff
@@ -84,79 +84,6 @@ paths:
         '404':
           $ref: '#/responses/resourceNotFound'
 
-  /api/v1/releases/{release}/revisions:
```
Yeah, I can see an argument that. As far as I know though, nobody looks at releases history through the public API though, perhaps we can defer this until it becomes needed?

nthomas-mozilla replied:
Looks like a whopping 7 requests for that endpoint since 2019-03-07, compared to 56K beginning with /api/v1/releases, so that seems fair to me. That's in the balrog_prod.balrog_elb_logs_aus_api data.

mozbhearsum commented on auslib/db.py:
```diff
@@ -662,7 +672,51 @@ def getRecentChanges(self, limit=10, transaction=None):
+        timestamp = getMillisecondTimestamp()
+        identifier = columns.get(self.identifier_column)
+        for data_version, ts, data in ((None, timestamp - 1, ""), (columns.get("data_version"), timestamp, json.dumps(columns[self.data_column]))):
+            bname = "{}/{}-{}-{}.json".format(identifier, data_version, ts, changed_by)
```
It's always 1 right now AFAICT (because of https://github.com/mozilla/balrog/blob/a7437b21bd3576c2ec8f8680264d57b24f0cd8e1/auslib/db.py#L408), but since `AUSTable` is responsible for deciding it, I thought it best to grab it from the args rather than make that assumption.

nthomas-mozilla replied:
Ok, sounds good.

mozbhearsum commented on auslib/migrate/versions/033_drop_releases_history.py:
```diff
@@ -0,0 +1,38 @@
+def upgrade(migrate_engine):
+    metadata = MetaData(bind=migrate_engine)
+
+    Table("releases_history", metadata, autoload=True).drop()
```
I'm actually not sure what our backup situation for `releases_history` looks like, now that you mention it. Certainly simply the migration won't restore the data though, of course. @oremj - can you speak to this?

The other possibility is to land the new code without dropping the table (which make require splitting the patch, depending how easy/difficult it is to have a migration in the repo that hasn't been run yet).

mozbhearsum commented on auslib/test/admin/views/base.py:
```diff
@@ -59,7 +60,15 @@ def my_userinfo(*args, **kwargs):
+        # only the merge tests benefit from having history
+        # the rest are fine if we just let history be None
+        # maybe we should mock this only in the tests we care about?
+        dbo.releases.history = FakeGCSHistory()
```
Yeah...I'll make some improvements here.

mozbhearsum commented on auslib/db.py:
```diff
@@ -662,7 +672,51 @@ def getRecentChanges(self, limit=10, transaction=None):
+        identifier = rowData.get(self.identifier_column)
+        bname = "{}/{}-{}-{}.json".format(identifier, rowData.get("data_version"), getMillisecondTimestamp(), changed_by)
+        blob = self._getBucket(identifier).blob(bname)
+        blob.upload_from_string(json.dumps(rowData[self.data_column]), content_type="application/json")
```
Nope, that's still an issue after this patch. Every case I've of duplicate data versions has ended up with different timestamps though, so the possibility of data loss is extremely small AFAICT. The real fix for this is likely to happen in `Releases.update`.

nthomas-mozilla replied:
OK, the situation is improving despite an existing issue.

nthomas-mozilla commented:
> I suspect you weren't connected to the VPN. For the new GCP environments, this is the expected response (unlike AWS, where we didn't return any response at all).

OK, I bounced the VPN and it started working, despite the original session being younger than this PR, so `¯\_(ツ)_/¯`. Maybe a caching thing.

A few things I noticed:
* 'View Data' buttons in the releases list (eg `/releases#68.0b1-build19`) doesn't show pretty printed JSON, it's a long line
* the release history pages (eg `/releases/Firefox-68.0b1-build19`) doesn't have that problem, but the JSON keys are not sorted (eg locales aren't listed alphabetically)
* 'View Diff' on data version 30 of Firefox-68.0b1-build19 has a bigger diff than in reality because of the lack of sorting (just `he` is being added to win arm64 in that change)
* suspect that's all client-side JS, because the JSON from the backend looks OK
* nit, the throbber never stops if there is no history for a release
* 'Change Logs' can't search for release changes any more, not very convenient to do that now ?

Didn't try upload etc as no admin permissions set up.

mozbhearsum commented:
I just granted you admin permissions. I'll look into the issues you mentioned - all the diff/view stuff should be relatively easy to fix.

mozbhearsum commented:
> 'Change Logs' can't search for release changes any more, not very convenient to do that now ?

Because this view is so buggy, and this UI is going away in the near future, I didn't bother to worry about this. I do want to implement a better version of this is the new UI though, so we can worry about it there.

mozbhearsum commented:
I think I've addressed all of the issues you brought up, except the deployment/rollback one (which obviously can't be fixed with code). I will get a plan together for that before we rollout, though.

nthomas-mozilla commented:
I'll take another run at this first thing next week.

mozbhearsum commented:
> There seems to be some fallout from the CSP changes on master which would need fixing before landing. I see this in the browser console

Good catch, I'll fix that up.

> One nit which you can do with as you like - the tests selecting history entries using ... if name.startswith("dd"). I see the tests are all carefully setup so we don't have any history collisions despite having a names liked and dd, a and ab, etc. To avoid future problems the code could add the delimeter, eg ... if name.startswith("dd/").

Should be easy, may as well do this at the same time!
